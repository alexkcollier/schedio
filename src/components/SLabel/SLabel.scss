.label {
  @include font-size(5);
  display: block;
  font-weight: $font-weight-bold;
  text-transform: uppercase;

  &--inline {
    align-items: center;
    display: inline-flex;
    flex-direction: row;
    justify-content: flex-start;

    .input,
    .textarea,
    .select {
      margin-left: spacing();
    }

    &.label--reversed {
      flex-direction: row-reverse;
    }
  }

  &--fullwidth {
    width: 100%;
  }

  &:not(:last-child) {
    margin-right: spacing();
  }

  &:not(:first-child) {
    margin: spacing('loose') 0;
  }

  &__optional {
    @include font-size(6);
    color: color('night', 'lightest');
    font-weight: $font-weight-light;
    text-transform: none;
  }
}
