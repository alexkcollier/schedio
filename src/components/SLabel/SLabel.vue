<template>
  <label :class="classList" class="label">
    <!-- Wrapped to keep optional on the right when reversed -->
    <span>
      {{ label }}
      <span v-if="isOptional" class="label__optional">&mdash;&nbsp;Optional</span>
    </span>
    <slot/>
  </label>
</template>

<script>
export default {
  name: 'SLabel',

  props: {
    label: {
      type: String,
      required: true
    },

    isOptional: {
      type: Boolean,
      default: true
    },

    isInline: {
      type: Boolean,
      default: false
    },

    isFullwidth: {
      type: Boolean,
      default: false
    },

    isReversed: {
      type: Boolean,
      default: false
    }
  },

  computed: {
    classList() {
      return {
        'label--inline': this.isInline,
        'label--fullwidth': this.isFullwidth,
        'label--reversed': this.isReversed
      }
    }
  },

  mounted() {
    if ((this.isFullwidth || this.isReversed) && !this.isInline) {
      console.error('`isFullwidth` and `isReversed` require `isInline`')
    }
  }
}
</script>
