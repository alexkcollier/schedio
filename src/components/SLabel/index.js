import SLabel from './SLabel.vue'

export { SLabel }

export default { install: Vue => Vue.component(SLabel.name, SLabel) }
