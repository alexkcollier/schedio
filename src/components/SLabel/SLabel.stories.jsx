import { SHeading } from '@/components/SHeading'
import { SLabel } from '@/components/SLabel'
import PropList from '@@/docs/components/PropList'
import { boolean, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/vue'

storiesOf('Components/Inputs.SLabel', module)
  .addParameters({ jest: 'SLabel' })
  .add('Label', () => {
    return {
      props: {
        props: {
          default: () => ({
            label: text('Label text', 'Label text', 'Required Props'),
            isFullwidth: boolean('isFullwidth', false, 'Optional Props'),
            isInline: boolean('isInline', false, 'Optional Props'),
            isReversed: boolean('isReversed', false, 'Optional Props'),
            isRequired: boolean('iRrequired', false, 'Optional Props'),
            hideOptional: boolean('hide-optional', false, 'Optional Props')
          })
        },

        defaultSlot: {
          default: text('Default', 'SomeInputComponent', 'Slots')
        }
      },
      render(h) {
        const { props, defaultSlot } = this.$props
        return (
          <div>
            <SHeading level="1">Label</SHeading>
            <p>
              <code>SLabel</code> provides labels for all form inputs. It is included automatically
              for all input components in this library, and is designed to wrap components.
            </p>
            <p>
              By default, if a label is for an optional field, it will append '- Optional' to the
              label. This text can be hidden with the <code>hide-optional</code> prop.
            </p>

            <SHeading level="2">Example</SHeading>
            <SLabel {...{ props }}>
              <div style="text-transform: none; font-weight: 300;">
                <code>{'<' + defaultSlot + ' />'}</code>
              </div>
            </SLabel>

            <PropList component={SLabel} />
          </div>
        )
      }
    }
  })
