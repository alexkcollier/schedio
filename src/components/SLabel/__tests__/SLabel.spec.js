import { shallowMount } from '@vue/test-utils'
import SLabel from '@/components/SLabel/SLabel.vue'

describe('SLabel.vue', () => {
  const testBooleanProp = (propName, value = true) => {
    wrapper.setProps({ [propName]: value })
    expect(wrapper.html()).toMatchSnapshot()
  }

  const shallowMountWrapper = (additionalProps = {}) => {
    return shallowMount(SLabel, {
      propsData: { label: 'label', ...additionalProps }
    })
  }

  const errorSpy = jest.spyOn(global.console, 'error').mockImplementation(() => {})

  let wrapper

  beforeEach(() => {
    wrapper = shallowMountWrapper()
  })

  afterEach(() => {
    jest.resetAllMocks()
  })

  it('renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('requires label text', () => {
    shallowMount(SLabel)
    expect(errorSpy).toBeCalled()
  })

  it('can be required', () => {
    wrapper.setProps({ isOptional: false })
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('can be inline', () => {
    testBooleanProp('isInline')
  })

  it('can be fullwidth', () => {
    wrapper.setProps({ isInline: true })
    testBooleanProp('isFullwidth')
  })

  it('can be reversed', () => {
    wrapper.setProps({ isInline: true })
    testBooleanProp('isReversed')
  })

  it('validates `is-fullwidth` and `is-reversed` are used with `is-inline`', () => {
    shallowMountWrapper({ isFullwidth: true })
    expect(errorSpy).toBeCalled()

    shallowMountWrapper({ isReversed: true })
    expect(errorSpy).toBeCalledTimes(2)
  })
})
