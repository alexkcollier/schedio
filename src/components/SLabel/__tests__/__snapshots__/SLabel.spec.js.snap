// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SLabel.vue can be fullwidth 1`] = `
<label class="label label--inline label--fullwidth"><span>
    label
    <span class="label__optional">
      —&nbsp;Optional
    </span></span> </label>
`;

exports[`SLabel.vue can be inline 1`] = `
<label class="label label--inline"><span>
    label
    <span class="label__optional">
      —&nbsp;Optional
    </span></span> </label>
`;

exports[`SLabel.vue can be required 1`] = `
<label class="label"><span>
    label
    <span class="label__optional">
      —&nbsp;Optional
    </span></span> </label>
`;

exports[`SLabel.vue can be reversed 1`] = `
<label class="label label--inline label--reversed"><span>
    label
    <span class="label__optional">
      —&nbsp;Optional
    </span></span> </label>
`;

exports[`SLabel.vue can hide the optional tag 1`] = `
<label class="label"><span>
    label
    <!----></span> </label>
`;

exports[`SLabel.vue renders correctly 1`] = `
<label class="label"><span>
    label
    <span class="label__optional">
      —&nbsp;Optional
    </span></span> </label>
`;
